File: src/extra/img_hash_async.cpp
#include "img_hash_async.h"
#include "core/types.h"
#include "core/vec.hpp"

// Asynchronous functions for Image Hashing

CvStatus *pHashCompute_Async(Mat inputArr, CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::img_hash::pHash(*inputArr.ptr, dst);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *pHashCompare_Async(Mat a, Mat b, CvCallback_1 callback) {
  BEGIN_WRAP
  double result = cv::img_hash::PHash::create()->compare(*a.ptr, *b.ptr);
  callback(new double(result));
  END_WRAP
}

CvStatus *averageHashCompute_Async(Mat inputArr, CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::img_hash::averageHash(*inputArr.ptr, dst);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *averageHashCompare_Async(Mat a, Mat b, CvCallback_1 callback) {
  BEGIN_WRAP
  double result = cv::img_hash::AverageHash::create()->compare(*a.ptr, *b.ptr);
  callback(new double(result));
  END_WRAP
}

CvStatus *BlockMeanHash_Create_Async(int mode, CvCallback_1 callback) {
  BEGIN_WRAP
  auto hash = new cv::Ptr<cv::img_hash::BlockMeanHash>(cv::img_hash::BlockMeanHash::create(mode));
  callback(new BlockMeanHash{hash});
  END_WRAP
}

CvStatus *BlockMeanHash_GetMean_Async(BlockMeanHash self, CvCallback_1 callback) {
  BEGIN_WRAP
  auto vec = self.ptr->get()->getMean();
  callback(vecdouble_cpp2c_p(vec));
  END_WRAP
}

CvStatus *BlockMeanHash_SetMode_Async(BlockMeanHash self, int mode, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setMode(mode);
  callback();
  END_WRAP
}

CvStatus *BlockMeanHash_Compute_Async(BlockMeanHash self, Mat inputArr, CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  self.ptr->get()->compute(*inputArr.ptr, dst);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *
BlockMeanHash_Compare_Async(BlockMeanHash self, Mat hashOne, Mat hashTwo, CvCallback_1 callback) {
  BEGIN_WRAP
  double result = self.ptr->get()->compare(*hashOne.ptr, *hashTwo.ptr);
  callback(new double(result));
  END_WRAP
}

CvStatus *colorMomentHashCompute_Async(Mat inputArr, CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::img_hash::colorMomentHash(*inputArr.ptr, dst);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *colorMomentHashCompare_Async(Mat a, Mat b, CvCallback_1 callback) {
  BEGIN_WRAP
  double result = cv::img_hash::ColorMomentHash::create()->compare(*a.ptr, *b.ptr);
  callback(new double(result));
  END_WRAP
}

CvStatus *
marrHildrethHashCompute_Async(Mat inputArr, float alpha, float scale, CvCallback_1 callback) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::img_hash::marrHildrethHash(*inputArr.ptr, dst, alpha, scale);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *
marrHildrethHashCompare_Async(Mat a, Mat b, float alpha, float scale, CvCallback_1 callback) {
  BEGIN_WRAP
  double result = cv::img_hash::MarrHildrethHash::create(alpha, scale)->compare(*a.ptr, *b.ptr);
  callback(new double(result));
  END_WRAP
}

CvStatus *radialVarianceHashCompute_Async(
    Mat inputArr, double sigma, int numOfAngleLine, CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Mat dst;
  cv::img_hash::radialVarianceHash(*inputArr.ptr, dst, sigma, numOfAngleLine);
  callback(new Mat{new cv::Mat(dst)});
  END_WRAP
}

CvStatus *radialVarianceHashCompare_Async(
    Mat a, Mat b, double sigma, int numOfAngleLine, CvCallback_1 callback
) {
  BEGIN_WRAP
  double result =
      cv::img_hash::RadialVarianceHash::create(sigma, numOfAngleLine)->compare(*a.ptr, *b.ptr);
  callback(new double(result));
  END_WRAP
}

File: lib/src/contrib/img_hash_async.dart
library cv;

import 'dart:ffi' as ffi;
import 'dart:typed_data';
import 'package:ffi/ffi.dart';
import '../core/base.dart';
import '../core/mat.dart';
import '../opencv.g.dart' as cvg;
import 'img_hash.dart';

extension PHashAsync on PHash {
  Future<double> compareAsync(InputArray hashOne, InputArray hashTwo) async {
    final rval = await cvRunAsync<double>(
        (callback) => CFFI.pHashCompare_Async(hashOne.ref, hashTwo.ref, callback), (c, p) {
      final rval = p.cast<ffi.Double>().value;
      calloc.free(p);
      return c.complete(rval);
    });
    return rval;
  }

  Future<void> computeAsync(InputArray inputArr, OutputArray outputArr) async {
    await cvRunAsync0<void>(
        (callback) => CFFI.pHashCompute_Async(inputArr.ref, outputArr.ref, callback), (c) => c.complete());
  }
}

extension AverageHashAsync on AverageHash {
  Future<double> compareAsync(InputArray hashOne, InputArray hashTwo) async {
    final rval = await cvRunAsync<double>(
        (callback) => CFFI.averageHashCompare_Async(hashOne.ref, hashTwo.ref, callback), (c, p) {
      final rval = p.cast<ffi.Double>().value;
      calloc.free(p);
      return c.complete(rval);
    });
    return rval;
  }

  Future<void> computeAsync(InputArray inputArr, OutputArray outputArr) async {
    await cvRunAsync0<void>(
        (callback) => CFFI.averageHashCompute_Async(inputArr.ref, outputArr.ref, callback),
        (c) => c.complete());
  }
}

extension BlockMeanHashAsync on BlockMeanHash {
  Future<double> compareAsync(InputArray hashOne, InputArray hashTwo, [int? mode_]) async {
    mode_ ??= mode;
    final rval = await cvRunAsync<double>(
        (callback) => CFFI.BlockMeanHash_Compare_Async(ref, hashOne.ref, hashTwo.ref, callback), (c, p) {
      final rval = p.cast<ffi.Double>().value;
      calloc.free(p);
      return c.complete(rval);
    });
    return rval;
  }

  Future<double> compareSAsync(InputArray hashOne, InputArray hashTwo, [int mode = 0]) async {
    final rval = await cvRunAsync<double>(
        (callback) => CFFI.BlockMeanHash_Compare_Async(ref, hashOne.ref, hashTwo.ref, callback), (c, p) {
      final rval = p.cast<ffi.Double>().value;
      calloc.free(p);
      return c.complete(rval);
    });
    return rval;
  }

  Future<void> computeAsync(InputArray inputArr, OutputArray outputArr, [int? mode_]) async {
    mode_ ??= mode;
    await cvRunAsync0<void>(
        (callback) => CFFI.BlockMeanHash_Compute_Async(ref, inputArr.ref, outputArr.ref, callback),
        (c) => c.complete());
  }

  Future<void> computeSAsync(InputArray inputArr, OutputArray outputArr, [int mode = 0]) async {
    await cvRunAsync0<void>(
        (callback) => CFFI.BlockMeanHash_Compute_Async(ref, inputArr.ref, outputArr.ref, callback),
        (c) => c.complete());
  }

  Future<List<double>> getMeanAsync() async {
    final rval = await cvRunAsync2<List<double>>(
        (callback) => CFFI.BlockMeanHash_GetMean_Async(ref, callback), (c, p, p2) {
      final ret = p.cast<ffi.Pointer<ffi.Double>>();
      final length = p2.cast<ffi.Int>().value;
      if (length == 0) return c.complete(List<double>.empty());
      return c.complete(List.generate(length, (i) => ret[i].value));
    });
    return rval;
  }
}

extension ColorMomentHashAsync on ColorMomentHash {
  Future<double> compareAsync(InputArray hashOne, InputArray hashTwo) async {
    final rval = await cvRunAsync<double>(
        (callback) => CFFI.colorMomentHashCompare_Async(hashOne.ref, hashTwo.ref, callback), (c, p) {
      final rval = p.cast<ffi.Double>().value;
      calloc.free(p);
      return c.complete(rval);
    });
    return rval;
  }

  Future<void> computeAsync(InputArray inputArr, OutputArray outputArr) async {
    await cvRunAsync0<void>(
        (callback) => CFFI.colorMomentHashCompute_Async(inputArr.ref, outputArr.ref, callback),
        (c) => c.complete());
  }
}

extension NewMarrHildrethHashAsync on NewMarrHildrethHash {
  Future<double> compareAsync(InputArray hashOne, InputArray hashTwo) async {
    final rval = await cvRunAsync<double>(
        (callback) => CFFI.marrHildrethHashCompare_Async(hashOne.ref, hashTwo.ref, alpha, scale, callback),
        (c, p) {
      final rval = p.cast<ffi.Double>().value;
      calloc.free(p);
      return c.complete(rval);
    });
    return rval;
  }

  Future<void> computeAsync(InputArray inputArr, OutputArray outputArr) async {
    await cvRunAsync0<void>(
        (callback) => CFFI.marrHildrethHashCompute_Async(inputArr.ref, outputArr.ref, alpha, scale, callback),
        (c) => c.complete());
  }
}

extension NewRadialVarianceHashAsync on NewRadialVarianceHash {
  Future<double> compareAsync(InputArray hashOne, InputArray hashTwo) async {
    final rval = await cvRunAsync<double>(
        (callback) =>
            CFFI.radialVarianceHashCompare_Async(hashOne.ref, hashTwo.ref, sigma, numOfAngleLine, callback),
        (c, p) {
      final rval = p.cast<ffi.Double>().value;
      calloc.free(p);
      return c.complete(rval);
    });
    return rval;
  }

  Future<void> computeAsync(InputArray inputArr, OutputArray outputArr) async {
    await cvRunAsync0<void>(
        (callback) => CFFI.radialVarianceHashCompute_Async(
            inputArr.ref, outputArr.ref, sigma, numOfAngleLine, callback),
        (c) => c.complete());
  }
}
